File: crates/phactory/api/src/crypto/eip712.rs
use std::convert::TryInto;

use super::{evm_ecdsa_recover, MessageType, SignatureVerifyError};

use ethers::{
    contract::{Eip712, EthAbiType},
    types::{transaction::eip712::Eip712, Bytes},
};
use sp_core::ecdsa::Public;

#[derive(Debug, Clone, Eip712, EthAbiType)]
#[eip712(
    name = "Phat Contract Query",
    version = "1",
    salt = "phala/phat-contract"
)]
#[allow(non_snake_case)]
pub struct PhatContractQuery {
    pub description: String,
    pub encodedQuery: Bytes,
}

impl PhatContractQuery {
    pub fn new(encoded_query: Bytes) -> Self {
        Self {
            description: "You are signing a query request that would be sent to a Phat Contract."
                .into(),
            encodedQuery: encoded_query,
        }
    }
}

#[derive(Debug, Clone, Eip712, EthAbiType)]
#[eip712(
    name = "Phat Query Certificate",
    version = "1",
    salt = "phala/phat-contract"
)]
#[allow(non_snake_case)]
pub struct IssueQueryCertificate {
    pub description: String,
    pub timeToLive: String,
    pub encodedCert: Bytes,
}

impl IssueQueryCertificate {
    pub fn new(cert_bytes: Bytes, ttl: u32) -> Self {
        Self {
            description: "You are signing a Certificate that can be used to query Phat Contracts using your identity without further prompts.".into(),
            timeToLive: format!("The Certificate will be valid till block {ttl}."),
            encodedCert: cert_bytes,
        }
    }
}

pub(crate) fn recover(
    pubkey: &[u8],
    signature: &[u8],
    msg: &[u8],
    msg_type: MessageType,
) -> Result<Public, SignatureVerifyError> {
    let signature = signature
        .try_into()
        .or(Err(SignatureVerifyError::InvalidSignature))?;
    let message: Bytes = msg.to_vec().into();
    let message_hash = match msg_type {
        MessageType::Certificate { ttl } => {
            IssueQueryCertificate::new(message, ttl).encode_eip712()
        }
        MessageType::ContractQuery => PhatContractQuery::new(message).encode_eip712(),
    }
    .or(Err(SignatureVerifyError::Eip712EncodingError))?;
    let recovered_pubkey = evm_ecdsa_recover(signature, message_hash)?;
    if recovered_pubkey.as_ref() != pubkey {
        return Err(SignatureVerifyError::InvalidSignature);
    }
    Ok(recovered_pubkey)
}

#[test]
fn signing_cert_works() {
    let message = b"Hello".to_vec();
    let pubkey =
        hex::decode("029df1e69b8b7c2da2efe0069dc141c2cec0317bf3fd135abaeb69ee33801f5970").unwrap();
    let mm_signature = hex::decode("885074f9b49de7bb73d23a801a69fb160f7cfef50cbf710c7aaff70c5581d7ab63bb7a80e4d8b3983498a9ffb10130fdcb98d5d5c49b7f84e3f99f194c26dcdf1b").unwrap();
    assert!(recover(
        &pubkey,
        &mm_signature,
        &message,
        MessageType::Certificate { ttl: 42 }
    )
    .is_ok());
}

#[test]
fn signing_query_works() {
    let message = b"Hello".to_vec();
    let pubkey =
        hex::decode("029df1e69b8b7c2da2efe0069dc141c2cec0317bf3fd135abaeb69ee33801f5970").unwrap();
    let mm_signature = hex::decode("b0ba4176ef624a71837c1a63e9c502c02314414883413913c126e05bfb3fda60474e0758b49c91647cd735c11d2a9575647509bdd2aef4eb5c5eba0019463a681c").unwrap();
    assert!(recover(&pubkey, &mm_signature, &message, MessageType::ContractQuery).is_ok());
}

File: frontend/packages/sdk/src/pruntime/eip712.ts
import { type ApiPromise } from '@polkadot/api'
import { ApiTypes, type SubmittableExtrinsic } from '@polkadot/api/types'
import { type U256, type U64 } from '@polkadot/types-codec'
import { hexToString, hexToU8a, stringToU8a, u8aToHex } from '@polkadot/util'
import { blake2AsU8a, decodeAddress, encodeAddress, keccak256AsU8a, secp256k1Compress } from '@polkadot/util-crypto'
import type { Account, Address, Hex, TestClient, WalletClient } from 'viem'
import { hashMessage, recoverPublicKey } from 'viem'
import { type signTypedData } from 'viem/wallet'
import { signMessage } from 'viem/wallet'

// keccak256(b"phala/phat-contract")
const SALT: Readonly<Hex> = '0x0ea813d1592526d672ea2576d7a07914cef2ca301b35c5eed941f7c897512a00'

type SignTypedDataInput = Parameters<typeof signTypedData>[1]

/**
 * Get compressed formatted ether address for a specified account via a Wallet Client.
 */
export async function etherAddressToCompressedPubkey(
  client: WalletClient | TestClient,
  account: Account,
  msg = 'Allows to access the pubkey address.'
) {
  const sign = await signMessage(client, { account, message: msg })
  const hash = hashMessage(msg)
  const recovered = await recoverPublicKey({ hash, signature: sign })
  const compressedPubkey = u8aToHex(secp256k1Compress(hexToU8a(recovered)))
  return compressedPubkey
}

export interface EtherAddressToSubstrateAddressOptions {
  SS58Prefix?: number
  msg?: string
}

/**
 * Convert an Ethereum address to a Substrate address.
 */
export async function etherAddressToSubstrateAddress(
  client: WalletClient,
  account: Account,
  { SS58Prefix = 30, msg }: EtherAddressToSubstrateAddressOptions = {}
) {
  const compressedPubkey = await etherAddressToCompressedPubkey(client, account, msg)
  const substratePubkey = encodeAddress(blake2AsU8a(hexToU8a(compressedPubkey)), SS58Prefix)
  return substratePubkey as Address
}

const EVM_ADDRESS_SUFFIX = stringToU8a('@evm_address')

/**
 * Convert an EVM public key (both compressed & uncompressed are supported) to a Substrate pubkey.
 */
export function evmPublicKeyToSubstratePubkey(hex: string) {
  const pubkey = hexToU8a(hex)
  if (pubkey.length === 65) {
    const h32 = keccak256AsU8a(pubkey.subarray(1))
    const h20 = h32.subarray(12)
    return new Uint8Array([...h20, ...EVM_ADDRESS_SUFFIX])
  } else if (pubkey.length === 33) {
    return blake2AsU8a(pubkey)
  } else {
    throw new Error('Invalid public key length.')
  }
}

export function substrateAddressToEvmAddress(address: string) {
  const substratePubkey = decodeAddress(address)
  if (substratePubkey.length !== 32) {
    throw new Error('Invalid public key length.')
  }
  if (substratePubkey.subarray(20).toString() !== EVM_ADDRESS_SUFFIX.toString()) {
    throw new Error('This Substrate Address is not mapped from an EVM Address.')
  }
  const h20 = substratePubkey.subarray(0, 20)
  return u8aToHex(h20)
}

export function createEip712StructedDataSignCertificate(
  account: Account,
  encodedCert: string,
  ttl: number
): SignTypedDataInput {
  return {
    domain: {
      name: 'Phat Query Certificate',
      version: '1',
      salt: SALT,
    },
    message: {
      description:
        'You are signing a Certificate that can be used to query Phat Contracts using your identity without further prompts.',
      timeToLive: `The Certificate will be valid till block ${ttl}.`,
      encodedCert,
    },
    primaryType: 'IssueQueryCertificate',
    types: {
      EIP712Domain: [
        { name: 'name', type: 'string' },
        { name: 'version', type: 'string' },
        { name: 'salt', type: 'bytes32' },
      ],
      IssueQueryCertificate: [
        { name: 'description', type: 'string' },
        { name: 'timeToLive', type: 'string' },
        { name: 'encodedCert', type: 'bytes' },
      ],
    },
    account,
  }
}

export function createEip712StructedDataSignQuery(account: Account, encodedQuery: string): SignTypedDataInput {
  return {
    domain: {
      name: 'Phat Contract Query',
      version: '1',
      salt: SALT,
    },
    message: {
      description: 'You are signing a query request that would be sent to a Phat Contract.',
      encodedQuery: encodedQuery,
    },
    primaryType: 'PhatContractQuery',
    types: {
      EIP712Domain: [
        { name: 'name', type: 'string' },
        { name: 'version', type: 'string' },
        { name: 'salt', type: 'bytes32' },
      ],
      PhatContractQuery: [
        { name: 'description', type: 'string' },
        { name: 'encodedQuery', type: 'bytes' },
      ],
    },
    account,
  }
}

export interface Eip712Domain {
  name: string
  version: string
  chainId: number
  verifyingContract: Address
}

export function createEip712Domain(api: ApiPromise): Eip712Domain {
  try {
    const name = hexToString(api.consts.evmAccountMapping.eip712Name.toString())
    const version = hexToString(api.consts.evmAccountMapping.eip712Version.toString())
    const chainId = (api.consts.evmAccountMapping.eip712ChainID as U256).toNumber()
    const verifyingContract = api.consts.evmAccountMapping.eip712VerifyingContractAddress.toString() as Address
    return {
      name,
      version,
      chainId,
      verifyingContract,
    }
  } catch (_err) {
    throw new Error(
      'Create Eip712Domain object failed, possibly due to the unavailability of the evmAccountMapping pallet.'
    )
  }
}

export interface SubstrateCall {
  who: string
  callData: string
  nonce: number
}

export async function createSubstrateCall<T extends ApiTypes>(
  api: ApiPromise,
  substrateAddress: string,
  extrinsic: SubmittableExtrinsic<T>
) {
  const nonce = await api.query.evmAccountMapping.accountNonce<U64>(substrateAddress)
  return {
    who: substrateAddress,
    callData: extrinsic.inner.toHex(),
    nonce: nonce.toNumber(),
  }
}

/**
 * @params account Account  The viem WalletAccount instance for signging.
 * @params who string       The SS58 formated address of the account.
 * @params callData string  The encoded call data, usually create with `api.tx.foo.bar.inner.toHex()`
 * @params nonce number     The nonce of the account.
 */
export function createEip712StructedDataSubstrateCall(
  account: Account,
  domain: Eip712Domain,
  message: SubstrateCall
): SignTypedDataInput {
  return {
    account,
    types: {
      EIP712Domain: [
        {
          name: 'name',
          type: 'string',
        },
        {
          name: 'version',
          type: 'string',
        },
        {
          name: 'chainId',
          type: 'uint256',
        },
        {
          name: 'verifyingContract',
          type: 'address',
        },
      ],
      SubstrateCall: [
        { name: 'who', type: 'string' },
        { name: 'callData', type: 'bytes' },
        { name: 'nonce', type: 'uint64' },
      ],
    },
    primaryType: 'SubstrateCall',
    domain: domain,
    message: { ...message },
  }
}
